### PR TITLE
change the contained class of predefined_macros_ from string_piece to…

### LINKED ISSUE #104: shaderc_compile_options_add_macro_definition doesn't copy macro name and value
The blurb for this function promises that name/value can be modified or deleted after it returns.  But that's currently not true: the underlying implementation merely keeps pointers to name/value.  The function should obey its contract.

### PR DESCRIPTION
**Pull request by Qining**
try to resolve #104

### CODE REVIEW

File: libshaderc_util/include/libshaderc_util/compiler.h [hunk 1]
```diff
@@ -76,8 +75,8 @@ class Compiler {
 
   // Adds an implicit macro definition obeyed by subsequent CompileShader()
   // calls.
-  void AddMacroDefinition(const string_piece& macro,
```
Comment (dneto0) [inline libshaderc_util/include/libshaderc_util/compiler.h]: Should say the second argument can be nullptr, and how that's interpreted.
Add a test for that case.

**Comment by dneto0** [comment]
-1 should update the blurb to describe nullability of the "definition" part of the value.

I feel we should be describing the const char* arguments as "null terminated strings".  The null terminated part matters.

Also, I'm getting uneasy with the plethora of interfaces that take a c-style string but no size argument.  Yes, it's less convenient, but forces the caller to think about sizing.

File: libshaderc_util/include/libshaderc_util/compiler.h (hunk 1 above)
Comment (dneto0) [inline libshaderc_util/include/libshaderc_util/compiler.h]: This comment is still unaddressed.
Update blurb, and add tests for the nullptr case.

**Comment by dneto0** [comment]
-1  Still unaddressed comments.

**Comment by Qining** [comment]
Added 'size' arguments to libshaderc and libshaderc_util and updated the blurbs.

Should we update glslc's code in this CL?

File: libshaderc_util/src/compiler_test.cc [hunk 2]
```diff
@@ -187,4 +187,21 @@ TEST_F(CompilerTest, DISABLED_RespectTargetEnvOnVulkanShader) {
   // TODO(dneto): Add test for a shader that should only compile for Vulkan.
 }
 
+TEST_F(CompilerTest, AddMacroDefinition) {
+  const std::string kMinimalExpandedShader = "void E(){}";
+  compiler_.AddMacroDefinition("E", 1u, "main", 4u);
```
Comment (dneto0) [inline libshaderc_util/src/compiler_test.cc]: Also add a test where the length matters for both arguments, i.e. the passed-in length is *smaller* than the length of the char* argument.

E.g. 
  AddMacroDefinition("EFGH", 1u, "mainly", 4u);

**Comment by dneto0** [comment]
-1
Need more test cases as described.

**Comment by Qining** [comment]
Add tests for 1) macro and value are not null terminated strings, 2) macro and value are null terminated, but the lengths of them passed in are too large.

File: libshaderc/include/shaderc.h [hunk 3]
```diff
@@ -136,16 +136,21 @@ shaderc_compile_options_t shaderc_compile_options_clone(
...
-// same effect as passing -Dname=value to the command-line compiler.
-// If value is NULL, it has the effect same as passing -Dname to the
-// command-line compiler. If a macro definition with the same name has
-// previously been added, the value is replaced with the new value.
-// The null-terminated strings that the name and value parameters point to
-// must remain valid for the duration of the call, but can be modified or
-// deleted after this function has returned.
+// Adds a predefined macro to the compilation options. This has the same
+// effect as passing -Dname=value to the command-line compiler.  If value
+// is NULL, it has the effect same as passing -Dname to the command-line
```
Comment (dekimir) [inline libshaderc/include/shaderc.h]: "effect same" => "same effect"

File: libshaderc/include/shaderc.h [hunk 4]
```diff
@@ -136,16 +136,21 @@ shaderc_compile_options_t shaderc_compile_options_clone(
...
+// compiler. If a macro definition with the same name has previously been
+// added, the value is replaced with the new value. The macro name and
+// value are passed in with char pointers, which point to their data, and
+// the lengths of their data. The strings that the name and value pointers
+// point to must remain valid for the duration of the call, but can be
+// modified or deleted after this function has returned. In case of adding
+// a valueless macro, the value argument can either be a null pointer or
+// point to an empty string. For valueless macro, the value_length parameter
+// will be ignored.
```
Comment (dekimir) [inline libshaderc/include/shaderc.h]: I think this is trying to say that you can denote a valueless macro by making `value` null or `value_length` 0 (or both).  So, clearly, `value_length` isn't ignored.  And "point to an empty string" seems ambiguous.

File: libshaderc/include/shaderc.hpp [hunk 5]
```diff
@@ -109,21 +109,29 @@ class CompileOptions {
     other.options_ = nullptr;
   }
 
-  // Adds a predefined macro to the compilation options.
+  // Adds a predefined macro to the compilation options. The macro name and
+  // value are passed in with their char* pointers of their strings and their
```
Comment (dekimir) [inline libshaderc/include/shaderc.hpp]: Stale bait.  Just omit this and say it behaves the same as `shaderc_compile_options_add_macro_definition`.

File: libshaderc/include/shaderc.hpp [hunk 6]
```diff
@@ -109,21 +109,29 @@ class CompileOptions {
...
+  // lengths. And the strings must remain valid for the duration of the call,
+  // but can be modified or deleted after this function has returned. The macro
+  // value argument can either be a nullptr or point to an empty string in case
+  // of adding a valueless macro. In case of valueless macro, the length of the
+  // value string will be ignored.
   // For further details refer to shaderc_compile_options_add_macro_definition
```
Comment (dekimir) [inline libshaderc/include/shaderc.hpp]: In fact, all the details are (redundantly) spelled out above -- there are no "further details" in `shaderc.h` right now.

File: libshaderc/src/shaderc_test.cc [hunk 7]
```diff
@@ -318,6 +319,25 @@ TEST_F(CompileStringWithOptionsTest, MacroCompileOptions) {
                                  cloned_options.get()));
 }
 
+TEST_F(CompileStringWithOptionsTest, MacroCompileOptionsNotNullTerminated) {
+  ASSERT_NE(nullptr, compiler_.get_compiler_handle());
+  shaderc_compile_options_add_macro_definition(options_.get(), "EFGH", 1u, "mainnnnnn",
+                                               4u);
+  const std::string kMinimalExpandedShader = "void E(){}";
+  const std::string kMinimalDoubleExpandedShader = "F E(){}";
```
Comment (dekimir) [inline libshaderc/src/shaderc_test.cc]: DRY

File: libshaderc_util/src/compiler_test.cc [hunk 8]
```diff
@@ -187,4 +187,27 @@ TEST_F(CompilerTest, DISABLED_RespectTargetEnvOnVulkanShader) {
...
+}
+
+TEST_F(CompilerTest, AddValuelessMacroDefinition) {
+  const std::string kMinimalExpandedShader =
+      "#ifdef E\n"
+      "void main(){}\n"
+      "#else\n"
+      "#error\n"
+      "#endif";
+  compiler_.AddMacroDefinition("E", 1u, nullptr, 0u);
```
Comment (dekimir) [inline libshaderc_util/src/compiler_test.cc]: I wouldn't set them both to zero.  We should be able to handle either of them being zero.

**Comment by dneto0** [comment]
+1 
Would like @dekimir to weigh in.

File: libshaderc/include/shaderc.h [hunk 9]
```diff
@@ -136,16 +136,20 @@ shaderc_compile_options_t shaderc_compile_options_clone(
...
+// Adds a predefined macro to the compilation options. This has the same
+// effect as passing -Dname=value to the command-line compiler.  If value
+// is NULL, it has the same effect as passing -Dname to the command-line
+// compiler. If a macro definition with the same name has previously been
+// added, the value is replaced with the new value. The macro name and
+// value are passed in with char pointers, which point to their data, and
+// the lengths of their data. The strings that the name and value pointers
+// point to must remain valid for the duration of the call, but can be
+// modified or deleted after this function has returned. In case of adding
+// a valueless macro, the value argument can be a null pointer or an empty
```
Comment (dekimir) [inline libshaderc/include/shaderc.h]: I still think "an empty string" is ambiguous and should be dropped.

File: libshaderc/include/shaderc.hpp [hunk 10]
```diff
@@ -109,21 +109,24 @@ class CompileOptions {
     other.options_ = nullptr;
   }
 
-  // Adds a predefined macro to the compilation options.
+  // Adds a predefined macro to the compilation options. It behaves the same as
+  // shaderc_compile_options_add_macro_definition in shaderc.h.
   // For further details refer to shaderc_compile_options_add_macro_definition
   // in shaderc.h
```
Comment (dekimir) [inline libshaderc/include/shaderc.hpp]: Now the above two lines can be dropped.

File: libshaderc_util/src/compiler_test.cc [hunk 11]
```diff
@@ -187,4 +196,28 @@ TEST_F(CompilerTest, DISABLED_RespectTargetEnvOnVulkanShader) {
   // TODO(dneto): Add test for a shader that should only compile for Vulkan.
 }
 
+TEST_F(CompilerTest, AddMacroDefinition) {
+  const std::string kMinimalExpandedShader = "void E(){}";
+  compiler_.AddMacroDefinition("E", 1u, "main", 4u);
+  EXPECT_TRUE(SimpleCompilationSucceeds(kMinimalExpandedShader, EShLangVertex));
+}
+
+TEST_F(CompilerTest, AddValuelessMacroDefinitionNullPointer) {
```
Comment (dekimir) [inline libshaderc_util/src/compiler_test.cc]: I think this test and the next would be more valuable at the libshaderc level.  More coverage of the user-visible interface, less implementation-dependent.

**Comment by Qining** [comment]
Tests about 'not null-terminated macro string' and 'zero length macro value string' added to libshaderc.

Blurbs updated.

File: libshaderc/include/shaderc.h [hunk 12]
```diff
@@ -136,16 +136,20 @@ shaderc_compile_options_t shaderc_compile_options_clone(
...
+// Adds a predefined macro to the compilation options. This has the same
+// effect as passing -Dname=value to the command-line compiler.  If value
+// is NULL, it has the same effect as passing -Dname to the command-line
+// compiler. If a macro definition with the same name has previously been
+// added, the value is replaced with the new value. The macro name and
+// value are passed in with char pointers, which point to their data, and
+// the lengths of their data. The strings that the name and value pointers
+// point to must remain valid for the duration of the call, but can be
+// modified or deleted after this function has returned. In case of adding
+// a valueless macro, the value argument should be a null pointer and the
```
Comment (dekimir) [inline libshaderc/include/shaderc.h]: "and" => "or"

Comment (Qining) [inline libshaderc/include/shaderc.h]: Should has been fixed by now.

**Comment by dekimir** [comment]
+2

**Comment by Qining** [comment]
Rebase and merged to master: 3bf8a0c243479f25afc0c3fcbde221a51edf1abd
